csrc/fusion_segmenter.cpp

```diff
@@ -3586,6 +3586,210 @@ bool CombineReductions::shouldRun(
+      // group. Also stop if there are multiple consumer edges to
+      // simplify keeping the DAG property.
+      if (!isDownCast(group) && group->consumer_edges.size() == 1) {
+        auto consumer_edge = group->consumer_edges.at(0);
+        SegmentedGroup* consumer_group = consumer_edge->to;
+        if (!groups_to_merge_set.count(consumer_group)) {
+          to_visit.push_back(consumer_group);
+        }
+      }
+
+      if (!isUpCast(group)) {
+        for (const auto producer_edge : group->producer_edges) {
+          SegmentedGroup* producer_group = producer_edge->from;
+          // Don't add producers that have more than multiple consumers
+          if (producer_group->consumer_edges.size() > 1) {
+            continue;
+          }
+          if (!groups_to_merge_set.count(producer_group)) {
+            to_visit.push_back(producer_group);
+          }
+        }
+      }
+    }
+
+    return groups_to_merge;
+  }
+
+  // Try merging a candidate cast group. Return a merged group if merged.
+  SegmentedGroup* mergeCastGroup(const std::vector<SegmentedGroup*>& groups) {
+    auto sched_type = tryMerge(
+        segment_candidate_finder_->segmented_fusion_.get(),
+        segment_candidate_finder_->runtimeInfo(),
+        groups);
+
+    if (sched_type == SchedulerType::None) {
+      return nullptr;
+    }
+
+    auto joined_group = segment_candidate_finder_->mergeAllGivenGroups(groups);
+
+    return joined_group;
+  }
+
+  bool isUpCast(SegmentedGroup* group) const {
+    if (auto precision_bits = getProducerConsumerPrecision(group);
+        precision_bits.has_value()) {
+      return precision_bits->first < precision_bits->second;
+    } else {
+      return false;
+    }
+  }
+
+  bool isDownCast(SegmentedGroup* group) const {
+    if (auto precision_bits = getProducerConsumerPrecision(group);
+        precision_bits.has_value()) {
+      return precision_bits->first > precision_bits->second;
+    } else {
+      return false;
+    }
+  }
+
+  std::optional<std::pair<int64_t, int64_t>> getProducerConsumerPrecision(
+      SegmentedGroup* group) const {
+    if (group->exprs().size() != 1) {
+      return std::nullopt;
+    }
+
+    auto uop = dynamic_cast<UnaryOp*>(group->exprs().front());
+    if (uop == nullptr || uop->getUnaryOpType() != UnaryOpType::Cast) {
+      return std::nullopt;
+    }
+
+    auto inp_tv = ir_utils::getTvInput(uop);
+    auto out_tv = ir_utils::getTvOutput(uop);
+    if (inp_tv == nullptr || out_tv == nullptr) {
+      return std::nullopt;
+    }
+
+    auto inp_dtype = inp_tv->dtype().type;
+    auto out_dtype = out_tv->dtype().type;
+    auto inp_prim_type = std::get_if<PrimDataType>(&inp_dtype);
+    auto out_prim_type = std::get_if<PrimDataType>(&out_dtype);
+
+    if (inp_prim_type == nullptr || out_prim_type == nullptr) {
+      return std::nullopt;
+    }
+
+    return std::make_pair(
+        primDataTypeSize(*inp_prim_type), primDataTypeSize(*out_prim_type));
```

Comment:
I think you needed a bail out here when either `inp_prim_type` or `out_prim_type` are `PrimDataType::Index`, which `primDataTypeSize` would throw an exception. I think welford would have that.

I only know that when I hit it in the presegmentation passes. :)

Reply:
I feel we could use a test for that. We don't have to do it in this PR though.

Reply:
Thanks. Updated.

Reply:
Added a test